### ISSUE TITLE
Feature: more complex filtering on song history, and duration

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#503]
There are several ways this might be done, but here's what I need:  I want to know exactly which songs I called for <club name> on Wednesday nights between 7:30 and 9:30 (local time).

Current method: keep track on paper, double check against exported-to-CSV song list with grep, matching up the dates by hand against <club name>'s dates.  (I call for 3 clubs who all meet on Wednesday nights.)

What I want ideally:  a filter on the export or maybe a new, separate function, something like "Show filtered song play history", with filter by day-of-week and between(time1,time2), to get me a lot closer to "when did I actually use this song last with this club?".

Also right now the song history only has the local start time (and same time in UTC), but what I really need is "how long was this song actually played?".  If the answer is 3 minutes, then I most likely actually used it.  If the answer is 10 seconds, then I might have loaded it, played it, but changed my mind at the last second, and I don't want to count it.  So, this is also a request to put in the Duration of the song playback into the song history (that is non-trivial, I think).

It just takes too much time to track this by hand, so that I don't reuse songs over and over at the different clubs (but it is OK to use the same song at two different clubs).

My calling with these clubs is irregular, so the Sessions feature does not work for me.

**Comment by danlyke** [comment]
I'd also like a little editing capability, like "I had the 'don't
record plays' thing on, but did", or "I only played a little bit of
this and want to remove the play".

Does it deserve its own tab?

**Comment by mpogue2** [comment]
Where possible, I think we should rely on external programs (where they are available and appropriate).  So, I'd prefer to see something like "open song play list in external editor", rather than building in an editor.

In the case of lyrics editing, there is no external program that is available to do the job, so it falls to us to do something that is tailored to lyric editing.  In this case, however, I think creating a text or CSV file, and then opening the default app associated with that type of file might be better, since those editors are far better than whatever we could build in.

Maybe for what you want, it looks more like export to external app, and then import?  Even that might be too complex for most users (other than programmers like you and me), because then do we allow adding lines on the fly?  etc...complexity....

On the Mac there's a nice editor for editing song history, that edits the database directly.  I'm not sure what's available on other platforms.

Interesting idea...I can see several ways to do what you want, but it seems complex to me to make it easy enough for everybody to use.

**Comment by danlyke** [comment]
What would you use for editing song history? I mean, play lists are
fine, but history is in our DB... though maybe there's some SQLite
editor that we can give a little schema help to?

**Comment by mpogue2** [comment]
If I have to edit the DB (and don't get me wrong -- I try hard to not ever have to do that!), I use DB Browser for SQLite on Mac OS X.  Maybe it has a way to specify schema to make it easier to edit -- I don't know for sure -- if it does, I've never used that feature.  It's great for browsing and simple editing, though.

**Comment by mpogue2** [comment]
The main feature I keep needing in the short term is "how long did I play that song?".  The underlying question is "what songs did I play for ECR when I called there over the last year?".  Playing for 5 or 10 seconds doesn't count for this, it would need to be at least, oh, maybe 4 minutes minimum to count as "played that night"?  If the Export function can include the Duration of playback, then I can export to CSV and filter there to get my answer.

OK, next question you're going to ask is "what is Duration exactly?".  If I play for 5 sec, then wait 1 minute, and play another 5 sec, is that two playbacks of 5 sec each, or one playback of 10 sec, or one playback of 70 sec?  I guess we get to define that.

1) If we export that as 2 5-sec segments, I can consolidate segments externally using an R script.  But, I think we should do it internally, so everybody can do it in a single step.
2) I think that probably argues for either 2 5-sec segments (keep it simple!), or 1 70-sec segment.

Should loading a different song start a new playback segment?  e.g. play 5 sec of song A, then play 5 sec of song B, then 5 sec of song A.  Is that a segment of 5 (or 10 or 70) seconds that possibly overlaps in time with song B?  Or, is it just 3 separate playback segments?

Certainly pressing PERIOD should not start a new segment, although it moves the playback head.
PAUSE/RESUME probably should not create a new segment, because we're probably just giving instruction (and the song was still "used").

Maybe we should do this:  export dialog should default to "everything is a separate segment", with a "consolidate" option that will merge segments that are the same song played the same evening.  The start time would be the start time of the first consolidated segment, and the duration would be the sum of the segment times that were consolidated.  Default consolidation would be, say, 3 hours, that is, if the same song was played multiple times within 3 hours, it counts as one song after consolidation and the record is merged.

Hmmm...this question is more complex than I expected.

**Comment by danlyke** [comment]
yeah, as I was sitting on the airplane this morning I was thinking
"well, if I just write the play time to the last record when you hit
stop..."

Which might work, and then we could put a "suppress plays shorter
than..." in the preferences dialog.

This would be a quick and dirty way to let me play a few things while I
set up to see what I want to do during my gig.

Related: I like the idea of "let's view plays by sessions and be able
to move everything I played last Wednesday from my Napa session to my
Tam Twirlers session" that you've hinted at.

**Comment by mpogue2** [comment]
Epic is now #1721.
